// @kyvejs/sdk/dist/clients/lcd-client/query/v1beta1/query.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (g && (g = 0, op[0] && (_ = 0)), _) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.QueryModuleLCDClient = void 0;
var lcd_client_abstract_1 = require("../../lcd-client.abstract");
var QueryModuleLCDClient = /** @class */ (function (_super) {
    __extends(QueryModuleLCDClient, _super);
    function QueryModuleLCDClient(restEndpoint) {
        return _super.call(this, restEndpoint) || this;
    }
    QueryModuleLCDClient.prototype.params = function () {
        return __awaiter(this, void 0, void 0, function () {
            var endpoint;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        endpoint = "/kyve/query/v1beta1/params";
                        return [4 /*yield*/, this.request(endpoint)];
                    case 1: return [2 /*return*/, _a.sent()];
                }
            });
        });
    };
    /** Pools **/
    QueryModuleLCDClient.prototype.pool = function (params) {
        return __awaiter(this, void 0, void 0, function () {
            var endpoint;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        endpoint = "/kyve/query/v1beta1/pool/".concat(params.id);
                        return [4 /*yield*/, this.request(endpoint)];
                    case 1: return [2 /*return*/, _a.sent()];
                }
            });
        });
    };
    QueryModuleLCDClient.prototype.pools = function (params) {
        return __awaiter(this, void 0, void 0, function () {
            var parameters, endpoint;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        parameters = {};
                        if (typeof (params === null || params === void 0 ? void 0 : params.pagination) !== "undefined") {
                            parameters.pagination = params.pagination;
                        }
                        if (typeof (params === null || params === void 0 ? void 0 : params.search) !== "undefined") {
                            parameters.search = params.search;
                        }
                        if (typeof (params === null || params === void 0 ? void 0 : params.runtime) !== "undefined") {
                            parameters.runtime = params.runtime;
                        }
                        if (typeof (params === null || params === void 0 ? void 0 : params.disabled) !== "undefined") {
                            parameters.disabled = params.disabled;
                        }
                        if (typeof (params === null || params === void 0 ? void 0 : params.storage_provider_id) !== "undefined") {
                            parameters.storageProviderId = params.storage_provider_id;
                        }
                        endpoint = "/kyve/query/v1beta1/pools";
                        return [4 /*yield*/, this.request(endpoint, parameters)];
                    case 1: return [2 /*return*/, _a.sent()];
                }
            });
        });
    };
    /** end Pools **/
    /** Stakers **/
    QueryModuleLCDClient.prototype.stakers = function (params) {
        return __awaiter(this, void 0, void 0, function () {
            var parameters, endpoint;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        parameters = {};
                        if (typeof (params === null || params === void 0 ? void 0 : params.pagination) !== "undefined") {
                            parameters.pagination = params.pagination;
                        }
                        endpoint = "/kyve/query/v1beta1/stakers";
                        return [4 /*yield*/, this.request(endpoint, params)];
                    case 1: return [2 /*return*/, _a.sent()];
                }
            });
        });
    };
    QueryModuleLCDClient.prototype.staker = function (params) {
        return __awaiter(this, void 0, void 0, function () {
            var endpoint;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        endpoint = "/kyve/query/v1beta1/staker/".concat(params.address);
                        return [4 /*yield*/, this.request(endpoint)];
                    case 1: return [2 /*return*/, _a.sent()];
                }
            });
        });
    };
    QueryModuleLCDClient.prototype.stakersByPool = function (params) {
        return __awaiter(this, void 0, void 0, function () {
            var endpoint;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        endpoint = "/kyve/query/v1beta1/stakers_by_pool/".concat(params.pool_id);
                        return [4 /*yield*/, this.request(endpoint)];
                    case 1: return [2 /*return*/, _a.sent()];
                }
            });
        });
    };
    /** end stakers **/
    /** Bundles **/
    QueryModuleLCDClient.prototype.currentVoteStatus = function (params) {
        return __awaiter(this, void 0, void 0, function () {
            var endpoint;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        endpoint = "/kyve/query/v1beta1/current_vote_status/".concat(params.pool_id);
                        return [4 /*yield*/, this.request(endpoint)];
                    case 1: return [2 /*return*/, _a.sent()];
                }
            });
        });
    };
    QueryModuleLCDClient.prototype.canValidate = function (params) {
        return __awaiter(this, void 0, void 0, function () {
            var endpoint;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        endpoint = "/kyve/query/v1beta1/can_validate/".concat(params.pool_id, "/").concat(params.valaddress);
                        return [4 /*yield*/, this.request(endpoint)];
                    case 1: return [2 /*return*/, _a.sent()];
                }
            });
        });
    };
    QueryModuleLCDClient.prototype.canPropose = function (params) {
        return __awaiter(this, void 0, void 0, function () {
            var endpoint;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        endpoint = "/kyve/query/v1beta1/can_propose/".concat(params.pool_id, "/").concat(params.staker, "/").concat(params.proposer, "/").concat(params.from_index);
                        return [4 /*yield*/, this.request(endpoint)];
                    case 1: return [2 /*return*/, _a.sent()];
                }
            });
        });
    };
    QueryModuleLCDClient.prototype.canVote = function (params) {
        return __awaiter(this, void 0, void 0, function () {
            var endpoint;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        endpoint = "/kyve/query/v1beta1/can_vote/".concat(params.pool_id, "/").concat(params.staker, "/").concat(params.voter, "/").concat(params.storage_id);
                        return [4 /*yield*/, this.request(endpoint)];
                    case 1: return [2 /*return*/, _a.sent()];
                }
            });
        });
    };
    /** end Bundles **/
    /** Delegations **/
    QueryModuleLCDClient.prototype.delegator = function (params) {
        return __awaiter(this, void 0, void 0, function () {
            var endpoint;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        endpoint = "/kyve/query/v1beta1/delegator/".concat(params.staker, "/").concat(params.delegator);
                        return [4 /*yield*/, this.request(endpoint)];
                    case 1: return [2 /*return*/, _a.sent()];
                }
            });
        });
    };
    QueryModuleLCDClient.prototype.delegatorsByStaker = function (params) {
        return __awaiter(this, void 0, void 0, function () {
            var parameters, endpoint;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        parameters = {};
                        if (typeof (params === null || params === void 0 ? void 0 : params.pagination) !== "undefined") {
                            parameters.pagination = params.pagination;
                        }
                        endpoint = "/kyve/query/v1beta1/delegators_by_staker/".concat(params.staker);
                        return [4 /*yield*/, this.request(endpoint, parameters)];
                    case 1: return [2 /*return*/, _a.sent()];
                }
            });
        });
    };
    QueryModuleLCDClient.prototype.stakersByDelegator = function (params) {
        return __awaiter(this, void 0, void 0, function () {
            var parameters, endpoint;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        parameters = {};
                        if (typeof (params === null || params === void 0 ? void 0 : params.pagination) !== "undefined") {
                            parameters.pagination = params.pagination;
                        }
                        endpoint = "/kyve/query/v1beta1/stakers_by_delegator/".concat(params.delegator);
                        return [4 /*yield*/, this.request(endpoint, parameters)];
                    case 1: return [2 /*return*/, _a.sent()];
                }
            });
        });
    };
    /** end Delegations **/
    /*** Account **/
    QueryModuleLCDClient.prototype.accountAssets = function (params) {
        return __awaiter(this, void 0, void 0, function () {
            var endpoint;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        endpoint = "/kyve/query/v1beta1/account_assets/".concat(params.address);
                        return [4 /*yield*/, this.request(endpoint)];
                    case 1: return [2 /*return*/, _a.sent()];
                }
            });
        });
    };
    QueryModuleLCDClient.prototype.accountDelegationUnbondings = function (params) {
        return __awaiter(this, void 0, void 0, function () {
            var endpoint;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        endpoint = "/kyve/query/v1beta1/account_delegation_unbondings/".concat(params.address);
                        return [4 /*yield*/, this.request(endpoint)];
                    case 1: return [2 /*return*/, _a.sent()];
                }
            });
        });
    };
    QueryModuleLCDClient.prototype.accountFundedList = function (params) {
        return __awaiter(this, void 0, void 0, function () {
            var endpoint;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        endpoint = "/kyve/query/v1beta1/account_funded_list/".concat(params.address);
                        return [4 /*yield*/, this.request(endpoint)];
                    case 1: return [2 /*return*/, _a.sent()];
                }
            });
        });
    };
    QueryModuleLCDClient.prototype.accountRedelegation = function (params) {
        return __awaiter(this, void 0, void 0, function () {
            var endpoint;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        endpoint = "/kyve/query/v1beta1/account_redelegation/".concat(params.address);
                        return [4 /*yield*/, this.request(endpoint)];
                    case 1: return [2 /*return*/, _a.sent()];
                }
            });
        });
    };
    /** End Account **/
    /** Funders **/
    QueryModuleLCDClient.prototype.funder = function (params) {
        return __awaiter(this, void 0, void 0, function () {
            var endpoint;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        endpoint = "/kyve/query/v1beta1/funder/".concat(params.address);
                        return [4 /*yield*/, this.request(endpoint)];
                    case 1: return [2 /*return*/, _a.sent()];
                }
            });
        });
    };
    QueryModuleLCDClient.prototype.funders = function (params) {
        return __awaiter(this, void 0, void 0, function () {
            var parameters, endpoint;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        parameters = {};
                        if (typeof (params === null || params === void 0 ? void 0 : params.pagination) !== "undefined") {
                            parameters.pagination = params.pagination;
                        }
                        if (typeof (params === null || params === void 0 ? void 0 : params.search) !== "undefined") {
                            parameters.search = params.search;
                        }
                        endpoint = "/kyve/query/v1beta1/funders";
                        return [4 /*yield*/, this.request(endpoint, parameters)];
                    case 1: return [2 /*return*/, _a.sent()];
                }
            });
        });
    };
    QueryModuleLCDClient.prototype.fundings_by_funder = function (params) {
        return __awaiter(this, void 0, void 0, function () {
            var endpoint;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        endpoint = "/kyve/query/v1beta1/fundings_by_funder/".concat(params.address);
                        return [4 /*yield*/, this.request(endpoint, params)];
                    case 1: return [2 /*return*/, _a.sent()];
                }
            });
        });
    };
    QueryModuleLCDClient.prototype.fundings_by_pool = function (params) {
        return __awaiter(this, void 0, void 0, function () {
            var endpoint;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        endpoint = "/kyve/query/v1beta1/fundings_by_pool/".concat(params.pool_id);
                        return [4 /*yield*/, this.request(endpoint, params)];
                    case 1: return [2 /*return*/, _a.sent()];
                }
            });
        });
    };
    return QueryModuleLCDClient;
}(lcd_client_abstract_1.AbstractKyveLCDClient));
exports.QueryModuleLCDClient = QueryModuleLCDClient;
