// @kyvejs/sdk/dist/clients/rpc-client/kyve/stakers/v1beta1/stakers.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
Object.defineProperty(exports, "__esModule", { value: true });
var tx_registry_1 = require("../../../../../registry/tx.registry");
var signing_1 = require("../../../signing");
var KyveStakersMethods = /** @class */ (function (_super) {
    __extends(KyveStakersMethods, _super);
    function KyveStakersMethods() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    KyveStakersMethods.prototype.createStaker = function (value, options) {
        var _this = this;
        var tx = tx_registry_1.withTypeUrl.createStaker(__assign(__assign({}, value), { creator: this.account.address }));
        return new signing_1.PendingTx({ tx: [tx] }, function () {
            return _this.getPendingSignedTx(tx, options);
        });
    };
    KyveStakersMethods.prototype.updateMetadata = function (value, options) {
        var _this = this;
        var tx = tx_registry_1.withTypeUrl.updateMetadata(__assign(__assign({}, value), { creator: this.account.address }));
        return new signing_1.PendingTx({ tx: [tx] }, function () {
            return _this.getPendingSignedTx(tx, options);
        });
    };
    KyveStakersMethods.prototype.updateCommission = function (value, options) {
        var _this = this;
        var tx = tx_registry_1.withTypeUrl.updateCommission(__assign(__assign({}, value), { creator: this.account.address }));
        return new signing_1.PendingTx({ tx: [tx] }, function () {
            return _this.getPendingSignedTx(tx, options);
        });
    };
    KyveStakersMethods.prototype.claimCommissionRewards = function (value, options) {
        var _this = this;
        var tx = tx_registry_1.withTypeUrl.claimCommissionRewards(__assign(__assign({}, value), { creator: this.account.address }));
        return new signing_1.PendingTx({ tx: [tx] }, function () {
            return _this.getPendingSignedTx(tx, options);
        });
    };
    KyveStakersMethods.prototype.joinPool = function (value, options) {
        var _this = this;
        var tx = tx_registry_1.withTypeUrl.joinPool(__assign(__assign({}, value), { creator: this.account.address }));
        return new signing_1.PendingTx({ tx: [tx] }, function () {
            return _this.getPendingSignedTx(tx, options);
        });
    };
    KyveStakersMethods.prototype.leavePool = function (value, options) {
        var _this = this;
        var tx = tx_registry_1.withTypeUrl.leavePool(__assign(__assign({}, value), { creator: this.account.address }));
        return new signing_1.PendingTx({ tx: [tx] }, function () {
            return _this.getPendingSignedTx(tx, options);
        });
    };
    return KyveStakersMethods;
}(signing_1.KyveSigning));
exports.default = KyveStakersMethods;
