// @kyvejs/sdk/dist/clients/lcd-client/client.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.createKyveLCDClient = void 0;
var query_1 = require("./bundles/v1beta1/query");
var query_2 = require("./delegation/v1beta1/query");
var query_3 = require("./funders/v1beta1/query");
var query_4 = require("./global/v1beta1/query");
var query_5 = require("./pool/v1beta1/query");
var query_6 = require("./query/v1beta1/query");
var query_7 = require("./stakers/v1beta1/query");
var query_8 = require("./team/v1beta1/query");
var query_9 = require("./v1/bundles/query");
var KyveLCDClient = /** @class */ (function () {
    function KyveLCDClient(restEndpoint) {
        this.v1 = {
            bundles: new query_9.V1BundlesLCDClient(restEndpoint),
        };
        this.bundles = {
            v1beta1: new query_1.BundlesModuleLCDClient(restEndpoint),
        };
        this.delegation = {
            v1beta1: new query_2.DelegationModuleLCDClient(restEndpoint),
        };
        this.funders = {
            v1beta1: new query_3.FundersModuleLCDClient(restEndpoint),
        };
        this.global = {
            v1beta1: new query_4.GlobalModuleLCDClient(restEndpoint),
        };
        this.pool = {
            v1beta1: new query_5.PoolModuleLCDClient(restEndpoint),
        };
        this.query = {
            v1beta1: new query_6.QueryModuleLCDClient(restEndpoint),
        };
        this.stakers = {
            v1beta1: new query_7.StakersModuleLCDClient(restEndpoint),
        };
        this.team = {
            v1beta1: new query_8.TeamModuleLCDClient(restEndpoint),
        };
    }
    return KyveLCDClient;
}());
function createKyveLCDClient(restEndpoint) {
    return {
        kyve: new KyveLCDClient(restEndpoint),
    };
}
exports.createKyveLCDClient = createKyveLCDClient;
