// @kyvejs/sdk/dist/clients/rpc-client/signing.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (g && (g = 0, op[0] && (_ = 0)), _) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.KyveSigning = exports.PendingSignedTx = exports.PendingTx = void 0;
var crypto_1 = require("@cosmjs/crypto");
var encoding_1 = require("@cosmjs/encoding");
var stargate_1 = require("@cosmjs/stargate");
var tx_1 = require("cosmjs-types/cosmos/tx/v1beta1/tx");
var constants_1 = require("../../constants");
var PendingTx = /** @class */ (function () {
    function PendingTx(options, executor) {
        this.tx = options.tx;
        this.executor = executor;
    }
    PendingTx.prototype.then = function (onfulfilled, onrejected) {
        return this.executor().then(onfulfilled, onrejected);
    };
    PendingTx.prototype.catch = function (onrejected) {
        return this.executor().catch(onrejected);
    };
    PendingTx.prototype.finally = function (onfinally) {
        return this.executor().finally(onfinally);
    };
    return PendingTx;
}());
exports.PendingTx = PendingTx;
var PendingSignedTx = /** @class */ (function () {
    function PendingSignedTx(nativeClient, tx) {
        this.nativeClient = nativeClient;
        this.txBytes = tx.txRawBytes;
        this.fee = tx.fee;
        this.tx = tx.tx;
        this.txHash = (0, encoding_1.toHex)((0, crypto_1.sha256)(this.txBytes)).toUpperCase();
    }
    PendingSignedTx.prototype.execute = function () {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, this.nativeClient.broadcastTx(this.txBytes)];
                    case 1: return [2 /*return*/, _a.sent()];
                }
            });
        });
    };
    return PendingSignedTx;
}());
exports.PendingSignedTx = PendingSignedTx;
var KyveSigning = /** @class */ (function () {
    function KyveSigning(client, account, config) {
        this.nativeClient = client;
        this.account = account;
        this.config = config;
    }
    KyveSigning.prototype.getPendingSignedTx = function (tx, options) {
        return __awaiter(this, void 0, void 0, function () {
            var _a, _b;
            return __generator(this, function (_c) {
                switch (_c.label) {
                    case 0:
                        _a = PendingSignedTx.bind;
                        _b = [void 0, this.nativeClient];
                        return [4 /*yield*/, this.signTx(tx, options)];
                    case 1: return [2 /*return*/, new (_a.apply(PendingSignedTx, _b.concat([_c.sent()])))()];
                }
            });
        });
    };
    KyveSigning.prototype.signTx = function (tx, options) {
        return __awaiter(this, void 0, void 0, function () {
            var txArr, gasEstimation, multiplier, fee, txRaw, gasEstimation, multiplier, fee, txRaw, txRaw;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        txArr = Array.isArray(tx) ? tx : [tx];
                        if (!!options) return [3 /*break*/, 3];
                        return [4 /*yield*/, this.nativeClient.simulate(this.account.address, txArr, undefined)];
                    case 1:
                        gasEstimation = _a.sent();
                        multiplier = constants_1.GAS_MULTIPLIER;
                        fee = (0, stargate_1.calculateFee)(Math.round(gasEstimation * multiplier), stargate_1.GasPrice.fromString("".concat(this.config.gasPrice).concat(this.config.coinDenom)));
                        return [4 /*yield*/, this.nativeClient.sign(this.account.address, txArr, fee, "")];
                    case 2:
                        txRaw = _a.sent();
                        return [2 /*return*/, {
                                txRawBytes: tx_1.TxRaw.encode(txRaw).finish(),
                                fee: fee,
                                tx: txArr,
                            }];
                    case 3:
                        if (!(options.fee === "auto" || typeof options.fee === "number")) return [3 /*break*/, 6];
                        return [4 /*yield*/, this.nativeClient.simulate(this.account.address, txArr, options.memo)];
                    case 4:
                        gasEstimation = _a.sent();
                        multiplier = typeof options.fee === "number" ? options.fee : constants_1.GAS_MULTIPLIER;
                        fee = (0, stargate_1.calculateFee)(Math.round(gasEstimation * multiplier), stargate_1.GasPrice.fromString("".concat(this.config.gasPrice).concat(this.config.coinDenom)));
                        return [4 /*yield*/, this.nativeClient.sign(this.account.address, txArr, fee, options.memo || "")];
                    case 5:
                        txRaw = _a.sent();
                        return [2 /*return*/, {
                                txRawBytes: tx_1.TxRaw.encode(txRaw).finish(),
                                fee: fee,
                                tx: txArr,
                            }];
                    case 6: return [4 /*yield*/, this.nativeClient.sign(this.account.address, txArr, options.fee, options.memo || "")];
                    case 7:
                        txRaw = _a.sent();
                        return [2 /*return*/, {
                                txRawBytes: tx_1.TxRaw.encode(txRaw).finish(),
                                tx: txArr,
                                fee: options.fee,
                            }];
                }
            });
        });
    };
    return KyveSigning;
}());
exports.KyveSigning = KyveSigning;
