// @kyvejs/sdk/dist/constants.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.KYVE_LEAP_CONFIG = exports.KYVE_COSMOSTATION_CONFIG = exports.KYVE_KEPLR_CONFIG = exports.SUPPORTED_WALLETS = exports.SUPPORTED_CHAIN_CONFIGS = exports.COIN_TYPE = exports.GAS_MULTIPLIER = exports.GOV_AUTHORITY = exports.PREFIX = void 0;
var cosmos_1 = require("@keplr-wallet/cosmos");
exports.PREFIX = "kyve";
exports.GOV_AUTHORITY = "kyve10d07y265gmmuvt4z0w9aw880jnsr700jdv7nah";
exports.GAS_MULTIPLIER = 1.5;
exports.COIN_TYPE = 118;
// chains are always identified by their chain id
exports.SUPPORTED_CHAIN_CONFIGS = {
    "kyve-1": {
        chainId: "kyve-1",
        chainName: "KYVE",
        rpc: "https://rpc.kyve.network",
        rest: "https://api.kyve.network",
        coin: "KYVE",
        coinDenom: "ukyve",
        coinDecimals: 6,
        gasPrice: 0.02,
    },
    "kaon-1": {
        chainId: "kaon-1",
        chainName: "KYVE Kaon",
        rpc: "https://rpc.kaon.kyve.network",
        rest: "https://api.kaon.kyve.network",
        coin: "KYVE",
        coinDenom: "tkyve",
        coinDecimals: 6,
        gasPrice: 0.02,
    },
    "korellia-2": {
        chainId: "korellia-2",
        chainName: "KYVE Korellia",
        rpc: "https://rpc.korellia.kyve.network",
        rest: "https://api.korellia.kyve.network",
        coin: "KYVE",
        coinDenom: "tkyve",
        coinDecimals: 6,
        gasPrice: 0.02,
    },
    "kyve-beta": {
        chainId: "kyve-beta",
        chainName: "KYVE-Beta",
        rpc: "https://rpc.beta.kyve.network",
        rest: "https://api.beta.kyve.network",
        coin: "KYVE",
        coinDenom: "tkyve",
        coinDecimals: 6,
        gasPrice: 0.02,
    },
    "kyve-alpha": {
        chainId: "kyve-alpha",
        chainName: "KYVE Alpha",
        rpc: "https://rpc.alpha.kyve.network",
        rest: "https://api.alpha.kyve.network",
        coin: "KYVE",
        coinDenom: "tkyve",
        coinDecimals: 6,
        gasPrice: 0.02,
    },
    "kyve-local": {
        chainId: "kyve-local",
        chainName: "KYVE Local",
        rpc: "http://0.0.0.0:26657",
        rest: "http://0.0.0.0:1317",
        coin: "KYVE",
        coinDenom: "tkyve",
        coinDecimals: 6,
        gasPrice: 0.02,
    },
};
exports.SUPPORTED_WALLETS = {
    KEPLR: "KEPLR",
    COSMOSTATION: "COSMOSTATION",
    LEAP: "LEAP",
};
exports.KYVE_KEPLR_CONFIG = {
    rpc: "",
    rest: "",
    chainId: "",
    chainName: "",
    stakeCurrency: {
        coinDenom: "",
        coinMinimalDenom: "",
        coinDecimals: 0,
    },
    bip44: {
        coinType: exports.COIN_TYPE,
    },
    bech32Config: cosmos_1.Bech32Address.defaultBech32Config("kyve"),
    currencies: [],
    feeCurrencies: [],
    features: ["ibc-transfer", "ibc-go"],
};
exports.KYVE_COSMOSTATION_CONFIG = {
    chainId: "",
    restURL: "",
    chainName: "",
    baseDenom: "",
    displayDenom: "",
    addressPrefix: exports.PREFIX,
    coinType: exports.COIN_TYPE.toString(),
    decimals: 0,
    gasRate: {
        tiny: "1",
        low: "1.5",
        average: "3",
    },
};
exports.KYVE_LEAP_CONFIG = {
    rpc: "",
    rest: "",
    chainId: "",
    chainName: "",
    stakeCurrency: {
        coinDenom: "",
        coinMinimalDenom: "",
        coinDecimals: 0,
    },
    bip44: {
        coinType: 118,
    },
    bech32Config: cosmos_1.Bech32Address.defaultBech32Config("kyve"),
    currencies: [],
    feeCurrencies: [],
    features: ["ibc-transfer", "ibc-go"],
};
